daemon/daemon.go

```diff
@@ -148,28 +148,35 @@ func (daemon *Daemon) Install(eng *engine.Engine) error {
 	return nil
 }
 
-// Get looks for a container by the specified ID or name, and returns it.
-// If the container is not found, or if an error occurs, nil is returned.
-func (daemon *Daemon) Get(name string) *Container {
-	id, err := daemon.idIndex.Get(name)
-	if err == nil {
-		return daemon.containers.Get(id)
-	}
+// Get looks for a container with the provided prefix
+func (daemon *Daemon) Get(prefix string) (*Container, error) {
+	if containerByID := daemon.containers.Get(prefix); containerByID != nil {
 
-	if c, _ := daemon.GetByName(name); c != nil {
-		return c
+		// prefix is an exact match to a full container ID
+		return containerByID, nil
 	}
 
-	if err == truncindex.ErrDuplicateID {
-		log.Errorf("Short ID %s is ambiguous: please retry with more characters or use the full ID.\n", name)
+	containerByName, _ := daemon.GetByName(prefix)
+	containerId, indexError := daemon.idIndex.Get(prefix)
+
+	if containerByName != nil {
```

Comment:
@ewindisch I haven't seen any version of daemon.Get() that tried to do partial name matching (unless there was something quite a while ago in early 2014 or 2013).  What changes with this patch is that exact match on name "db" will trump partial **prefix** match "db" on a container with ID = "db123abf...".  In today's code the named container "db" is totally inaccessible until no containers that being with prefix "db.." are removed from the graph/index.

Simplified, the order of search is now:
- full hex ID match - always wins
- partial ID match wins unless partial ID == exact match on a name
- if no ID match at all, check exact name match
- no container found if we get here without a container result

As a comparison, today's code:
- full hex ID or partial "prefix" ID wins always
- if no full or partial ID match, then search names for exact name match
- if no (full or partial) ID match and no exact names match, then return no container found

Another glitch of today's code that is fixed in this PR is that "prefix ambiguous" is currently returned as no container match rather than "more than 1 match, please be more specific".